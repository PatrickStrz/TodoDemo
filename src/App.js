import React, { Component } from 'react'
import styled from 'styled-components'
import COLORS from './constants/Colors'
import STATUS from './constants/Status'
import CreateTodoForm from './components/CreateTodoForm'
import TodoItem from './components/TodoItem'
import Button from './components/Button'

const SiteBox = styled.div`
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  min-width: 100vw;
  justify-content: center;
  align-items: center;
  background-color: ${COLORS.background};
  /* ignore highlights in browser:  */
  input:focus,
  select:focus,
  textarea:focus,
  button:focus {
    outline: none;
  }
`

const FilterButtonsBox = styled.div`
  margin-bottom: 25px;
  display: flex;
  justify-content: space-between;
  align-items: center;
`

class App extends Component {
  state = {
    todos: [],
    editModalVisible: false,
    filterBy: null
  }
  render() {
    const getFilteredTodos = () => {
      switch (this.state.filterBy) {
        case null:
          return this.state.todos
        case STATUS.DONE:
          return this.state.todos.filter(todo => todo.status === STATUS.DONE)
        case STATUS.PENDING:
          return this.state.todos.filter(todo => todo.status === STATUS.PENDING)
      }
    }

    return (
      <SiteBox>
        <FilterButtonsBox>
          <Button
            active={this.state.filterBy === null}
            onClick={() => this._setFilter(null)}
          >
            ALL
          </Button>
          <Button
            active={this.state.filterBy === STATUS.PENDING}
            onClick={() => this._setFilter(STATUS.PENDING)}
          >
            PENDING
          </Button>
          <Button
            active={this.state.filterBy === STATUS.DONE}
            onClick={() => this._setFilter(STATUS.DONE)}
          >
            DONE
          </Button>
        </FilterButtonsBox>
        {getFilteredTodos().map(todo => (
          <TodoItem
            key={todo.id}
            id={todo.id}
            title={todo.title}
            description={todo.description}
            dueDate={todo.dueDate}
            onDeleteClick={this._handleDeleteClick}
            status={todo.status}
            onToggleStatus={this._handleToggleStatus}
            onEditClick={this._showEditModal}
            editModalVisible={this.state.editModalVisible}
            onEditModalClose={this._hideEditModal}
            onEditSubmit={this._handleEditSubmit}
          />
        ))}
        <CreateTodoForm onSubmit={this._handleTodoSubmit} />
      </SiteBox>
    )
  }

  _setFilter = filterName => this.setState({ filterBy: filterName })

  _addToList = (todos, newTodo) => {
    return [...todos, newTodo]
  }

  _handleTodoSubmit = todo => {
    this.setState((prevState, props) => {
      return { todos: this._addToList(prevState.todos, todo) }
    })
  }
  _handleDeleteClick = () => {
    alert('clicked delete')
  }

  _handleToggleStatus = id => {
    const { todos } = this.state
    const indexOfTodo = todos.findIndex(todo => todo.id === id)
    const todo = todos[indexOfTodo]
    const status = todo.status === STATUS.DONE ? STATUS.PENDING : STATUS.DONE
    const updatedTodo = {
      ...todo,
      status
    }
    const newTodos = [
      ...todos.slice(0, indexOfTodo),
      updatedTodo,
      ...todos.slice(indexOfTodo + 1)
    ]
    this.setState({ todos: newTodos })
  }

  _handleDeleteClick = id => {
    const { todos } = this.state
    const indexOfTodo = todos.findIndex(todo => todo.id === id)
    const newTodos = [
      ...todos.slice(0, indexOfTodo),
      ...todos.slice(indexOfTodo + 1)
    ]
    this.setState({ todos: newTodos })
  }

  _handleEditSubmit = (id, title, description, dueDate) => {
    const { todos } = this.state
    const indexOfTodo = todos.findIndex(todo => todo.id === id)
    const todo = todos[indexOfTodo]
    const updatedTodo = {
      ...todo,
      title,
      description,
      dueDate
    }
    const newTodos = [
      ...todos.slice(0, indexOfTodo),
      updatedTodo,
      ...todos.slice(indexOfTodo + 1)
    ]
    this.setState({ todos: newTodos })
  }

  _showEditModal = () => this.setState({ editModalVisible: true })
  _hideEditModal = () => this.setState({ editModalVisible: false })
}

export default App
