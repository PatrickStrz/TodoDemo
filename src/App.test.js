import React from 'react'
import ReactDOM from 'react-dom'
import App from './App'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

configure({ adapter: new Adapter() })

describe('<App/>', () => {
  it('should render App wrapper', () => {
    const app = shallow(<App />)
    console.log(app.debug())
  })
})

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<App />, div)
  ReactDOM.unmountComponentAtNode(div)
})
