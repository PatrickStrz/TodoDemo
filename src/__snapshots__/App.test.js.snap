// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> matches the snapshot 1`] = `
<styled.div>
  <styled.div>
    <styled.div
      active={true}
      onClick={[Function]}
    >
      ALL
    </styled.div>
    <styled.div
      active={false}
      onClick={[Function]}
    >
      PENDING
    </styled.div>
    <styled.div
      active={false}
      onClick={[Function]}
    >
      DONE
    </styled.div>
  </styled.div>
  <TodoItem
    description="description 1"
    dueDate=""
    editModalVisible={false}
    id="1"
    key="1"
    onDeleteClick={[Function]}
    onEditClick={[Function]}
    onEditModalClose={[Function]}
    onEditSubmit={[Function]}
    onToggleStatus={[Function]}
    status="PENDING"
    title="pending todo"
  />
  <TodoItem
    description="description 1"
    dueDate=""
    editModalVisible={false}
    id="2"
    key="2"
    onDeleteClick={[Function]}
    onEditClick={[Function]}
    onEditModalClose={[Function]}
    onEditSubmit={[Function]}
    onToggleStatus={[Function]}
    status="DONE"
    title="completed todo"
  />
  <TodoItem
    description="description 1"
    dueDate=""
    editModalVisible={false}
    id="3"
    key="3"
    onDeleteClick={[Function]}
    onEditClick={[Function]}
    onEditModalClose={[Function]}
    onEditSubmit={[Function]}
    onToggleStatus={[Function]}
    status="DONE"
    title="completed todo"
  />
  <CreateTodoForm
    onSubmit={[Function]}
  />
</styled.div>
`;
